layers: GH41 Flag error if cmd buffer references destroyed framebuffer

tobine opened the pull request:
Create framebuffer node and include set of cmd buffers that reference that FB.
At the time when RenderPass is bound to cmd buffer, or when secondary cmd buffer
is created with framebuffer inheritence, store a reference from FB to cmd buffer.

When FB destroyed, set any cmd buffers referencing it as INVALID and store FB ref.

At submit time, if CB is INVALID, check for any destroyed FBs and report them.

lentinem commented on layers/draw_state.cpp:
```diff
@@ -4104,21 +4105,48 @@ static bool validateCommandBufferState(layer_data *dev_data,
+                  reinterpret_cast<uint64_t&>(pCB->commandBuffer), __LINE__,
+                  DRAWSTATE_INVALID_COMMAND_BUFFER, "DS",
+                  "You are submitting command buffer %#" PRIxLEAST64
+                  " that is invalid due to an unknown cause. Validation should "
```
Why is this unknown? seems like it would be easy enough to make casueReported a enum that we can switch over and combine this with above.

tobine replied:
Currently there should be no cases where causeReported is false. I stewed over this code quite a bit and added this final block as a defensive mechanism in the event that we were missing any case where an INVALID cmd buffer was submitted, or if someone in the future causes such a case, I didn't want this code to be cleared without *some* error being flagged so I put this in both to make sure an error is flagged, but also to signal that validation should be updated to explicitly handle this case.

Using an enum here isn't ideal as there may be multiple cases that cause INVALID cmd buffer to be submitted so the separate "if" blocks are meant to report each of these cases.

I'm inclined to stick with the defensive code for now, although I don't like it a ton, and put in a TODO here to improve the INVALID cmd buffer reporting in general. We still have various holes that need to be filled here so it probably calls for a rework in the near future.

lentinem replied:
I would probably at least remove the else in this case (to be more consistent so that we have a bunch of if (...) {output} if (!output) ...) and add a comment to the affect that this should never happen. Did you see this happen in practice? I'm a bit of a fan of putting an assert here but realize that that might cause issues for users - depends how confident you are that it will never happen.

lentinem commented on layers/draw_state.cpp:
```diff
@@ -7040,8 +7105,12 @@ VK_LAYER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkCreateFramebuffer(VkDevice devi
         }
+        FRAMEBUFFER_NODE fbNode = {};
+        fbNode.createInfo = *localFBCI;
+        std::pair<VkFramebuffer, FRAMEBUFFER_NODE> fbPair(*pFramebuffer,
```
I would inline this with std::make_pair but am fine leaving it if you prefer this way.

lentinem commented on layers/draw_state.cpp:
```diff
@@ -7399,34 +7468,47 @@ VkBool32 VerifyFramebufferAndRenderPassLayouts(VkCommandBuffer cmdBuffer, const
-                                 "You cannot start a render pass using attachment %i where the intial layout differs from the starting layout.", i);
-                }
-            }
+      const VkImageView& image_view = framebufferInfo.pAttachments[i];
```
What was changed in this block?

tobine replied:
Interesting. I didn't change anything here and code looks the same so not sure why it was reformatted.

tobine replied:
Ah, now I see it. I changed "pFramebufferInfo->" to "framebufferInfo." git clang-format just reformatted a larger section than may have been expected.

lentinem commented on layers/draw_state.cpp:
```diff
@@ -7485,11 +7570,12 @@ void TransitionFinalSubpassLayouts(VkCommandBuffer cmdBuffer, const VkRenderPass
-        const VkImageView& image_view = pFramebufferInfo->pAttachments[i];
-        SetLayout(dev_data, pCB, image_view,
-                  pRenderPassInfo->pAttachments[i].finalLayout);
+      const VkImageView& image_view = framebufferInfo.pAttachments[i];
```
This looks wrong to me. I thought the indentation for clang-format was 4. Are you running with the correct file?

tobine replied:
I just ran "git clang-format" which I thought I had set up correctly but it's possible my config is not correct. Mark just said that he's planning to run all layers through clang-format today so if I get this in first, his change will override any errors and I can confirm separately that I have correct format.

lentinem replied:
Try git clang-format --style=file

lentinem commented on layers/draw_state.h:
```diff
@@ -669,10 +675,12 @@ typedef struct _GLOBAL_CB_NODE {
     // Track descriptor sets that are destroyed or updated while bound to CB
     std::set<VkDescriptorSet>    destroyedSets;
     std::set<VkDescriptorSet>    updatedSets;
+    unordered_set<VkFramebuffer> destroyedFramebuffers;
```
Does it make sense to combine this with destroyedSets and just have destroyedObjects or does knowing what type help?

tobine replied:
It makes sense to simplify this in some fashion, as there are still more cases of different resource types that can be deleted and cause INVALID cmd buffers, so I don't think we want separate sets for each. There's also a hole in this current fix where it doesn't handle FBs deleted out from under an in-flight cmd buffer.

I'd like to just consider this a spot-fix for a specific bug that was reported, and go ahead with the core code and perhaps a few added "TODOs". Then I'll file a separate bug to implement a more complete solution for handling deletion any resource that a cmd_buffer depends on. We do handle many of these cases already w/ the BASE_NODE and in_use tracking, but we need to build that out to handle missing cases. I'd like to get the mem_tracker merge done first, and then can address this issue fully.

lentinem replied:
I view the deleted out from under as part of fb lifetime which is on the list of valid usage tasks. The lifetime checks all have a common implementation (in_use) which I like and I would want something similar for the invalid cmd buffer submission checks like these so we can just add new objects into the checks more easily. I think this framework can be used but doing something like merging the sets will probably make it easier.

tobine commented:
I changed "else if" to "if", added a few "TODO" comments, correctly ran git clang-format and went ahead and pushed this to make SDK freeze. I'll open separate GH issue on myself to unify INVALID cmd_buffer handling.
